contrib/tools/workflowcheck/determinism/checker.go

```diff
@@ -0,0 +1,487 @@
+		skipFile := false
+		for _, skipPattern := range c.SkipFiles {
+			if skipFile = skipPattern.MatchString(fileName); skipFile {
+				break
+			}
+		}
+		if skipFile {
+			continue
+		}
+
+		// Update ignore map
+		updateIgnoreMap(pass.Fset, file, coll.ignoreMap)
+
+		// Collect the decls to check and check vars/iface patterns
+		for _, decl := range file.Decls {
+			switch decl := decl.(type) {
+			case *ast.FuncDecl:
+				// Collect top-level func
+				if funcType, _ := pass.TypesInfo.ObjectOf(decl.Name).(*types.Func); funcType != nil {
+					funcDecls[funcType] = decl
+				}
+			case *ast.GenDecl:
+				// Set top-level vars that match pattern as non-deterministic
+				for _, spec := range decl.Specs {
+					switch spec := spec.(type) {
+					// See if the top-level vars match patterns
+					case *ast.ValueSpec:
+						for _, varName := range spec.Names {
+							if varType, _ := pass.TypesInfo.ObjectOf(varName).(*types.Var); varType != nil && varType.Pkg() != nil {
+								fullName := varType.Pkg().Path() + "." + varType.Name()
+								if c.IdentRefs[fullName] {
+									c.debugf("Marking %v as non-deterministic because it matched a pattern", fullName)
+									pos := pass.Fset.Position(varType.Pos())
+									coll.nonDetVars[varType] = NonDeterminisms{&ReasonDecl{SourcePos: &pos}}
+								}
+							}
+						}
+					// See if any interface funcs match patterns
+					case *ast.TypeSpec:
+						if iface, _ := pass.TypesInfo.TypeOf(spec.Type).(*types.Interface); iface != nil {
+							// Only need to match explicitly defined methods
+							for i := 0; i < iface.NumExplicitMethods(); i++ {
+								info := coll.funcInfo(iface.ExplicitMethod(i))
+								if c.IdentRefs[info.fn.FullName()] {
+									c.debugf("Marking %v as non-deterministic because it matched a pattern", info.fn.FullName())
+									pos := pass.Fset.Position(spec.Pos())
+									info.reasons = append(info.reasons, &ReasonDecl{SourcePos: &pos})
+								}
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+
+	// Build collector and do initial pass for each function async
+
+	// Parallelize to the number of CPUs
+	maxAtOnce := runtime.NumCPU()
+	if maxAtOnce < 1 {
+		maxAtOnce = 1
+	}
+	doneCh := make(chan struct{}, maxAtOnce)
+	var running int
+	for fn, decl := range funcDecls {
+		// If we've filled the channel, wait
+		if running == cap(doneCh) {
+			<-doneCh
+		} else {
+			running++
+		}
+		go func(fn *types.Func, decl *ast.FuncDecl) {
+			coll.collectFuncInfo(fn, decl)
+			doneCh <- struct{}{}
+		}(fn, decl)
+	}
+	// Wait for the rest to finish
+	for i := 0; i < running; i++ {
+		<-doneCh
+	}
+	// Build facts in second pass
+	return coll.applyFacts(), nil
+}
+
+func updateIgnoreMap(fset *token.FileSet, f *ast.File, m map[ast.Node]struct{}) {
+	// Collect only the ignore comments
+	var comments []*ast.CommentGroup
+	for _, group := range f.Comments {
+		if len(group.List) == 1 && strings.HasPrefix(group.List[0].Text, "//workflowcheck:ignore") {
```

Comment:
We should probably allow whitespaces after `//`

Reply:
Disagree as this is what go would call a "directive" instead of a normal comment. Go traditionally differentiates the two by whether a space follows the `//`. Otherwise, if you put this above a function, it'd become part of its public documentation. See https://golangci-lint.run/usage/false-positives/ and https://staticcheck.io/docs/configuration/#ignoring-problems and others for inspiration.